src/persist-client/src/internal/machine.rs

```diff
@@ -578,16 +614,19 @@ where
 
                         return Ok((self.state.seqno(), Ok(work_ret), maintenance));
                     }
-                    Err(current) => {
+                    Err(live_diffs) => {
                         debug!(
                             "apply_unbatched_cmd {} {} lost the CaS race, retrying: {} vs {:?}",
                             self.shard_id(),
                             cmd.name,
                             self.state.seqno(),
-                            current.as_ref().map(|x| x.seqno)
+                            live_diffs.last().map(|x| x.seqno)
                         );
                         cas_mismatch_metric.0.inc();
-                        self.update_state(current).await;
+                        self.fetcher
+                            .update_state(&mut self.state, live_diffs)
```

Comment:
done! (mostly)

moved the apply methods to StateDiff instead of creating a new StateApplier. deconstructed update_state into roughly the pattern you mentioned (replacing Option<()> will bool 😂). still need some polish on fetch_latest_state, it's harder than you'd initially guess for reasons I'm too tired to explain today

Reply:
explained the difference between fetch_current_state and the new fetch_and_update_to_current in rustdoc (one is all live diffs, the other is only diffs between some in-mem state.seqno and current)